File: flowParser.js
/**
 * flowParser.js
 * Reads Salesforce Flow metadata from SFDX source format.
 */

'use strict';

const fs   = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

async function parseFlows(flowsDir) {
  const flows = [];

  if (!fs.existsSync(flowsDir)) return { flows };

  const files = fs.readdirSync(flowsDir).filter(f => f.endsWith('.flow-meta.xml'));

  for (const file of files) {
    const xml  = fs.readFileSync(path.join(flowsDir, file), 'utf8');
    const name = file.replace('.flow-meta.xml', '');
    flows.push({
      apiName:     name,
      label:       extractTag(xml, 'label'),
      processType: extractTag(xml, 'processType'),
      triggerType: extractTag(xml, 'triggerType'),
      object:      extractTag(xml, 'object'),
      status:      extractTag(xml, 'status'),
      description: extractTag(xml, 'description'),
    });
  }

  return { flows };
}

module.exports = { parseFlows };

File: scripts/python/parsers/flow_parser.py
"""
flow_parser.py
Reads Salesforce Flow metadata from SFDX source format.
Uses stdlib only (re, pathlib).
"""

import re
from pathlib import Path


def _extract_tag(xml: str, tag: str):
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def parse_flows(flows_dir: str) -> dict:
    """
    Parse all flows under flows_dir.
    Returns {'flows': [{'apiName', 'label', 'processType', 'triggerType', 'object', 'status', 'description'}]}
    """
    flows = []
    p = Path(flows_dir)

    if not p.exists():
        return {'flows': flows}

    for ff in sorted(p.glob('*.flow-meta.xml')):
        xml = ff.read_text(encoding='utf-8')
        name = ff.name.replace('.flow-meta.xml', '')
        flows.append({
            'apiName': name,
            'label': _extract_tag(xml, 'label'),
            'processType': _extract_tag(xml, 'processType'),
            'triggerType': _extract_tag(xml, 'triggerType'),
            'object': _extract_tag(xml, 'object'),
            'status': _extract_tag(xml, 'status'),
            'description': _extract_tag(xml, 'description'),
        })

    return {'flows': flows}
